pulpcore/app/viewsets/sam.py

```diff
@@ -0,0 +1,169 @@
+
+from pulpcore.plugin.tasking import dispatch
+
+
+class SharedAttributeManagerFilter(BaseFilterSet):
+    """
+    Filter for SharedAttributeManager.
+
+    You can look for SAMs by name, or by "manages this HREF".
+    """
+
+    name = filters.CharFilter()
+    managed_entities = ManagedEntitiesFilter()
+
+    class Meta:
+        model = SharedAttributeManager
+        fields = {
+            "name": ["contains"],
+            "managed_entities": ["exact"],
+        }
+
+
+class SharedAttributeManagerViewSet(
+    NamedModelViewSet,
+    mixins.CreateModelMixin,
+    mixins.RetrieveModelMixin,
+    mixins.ListModelMixin,
+    mixins.DestroyModelMixin,
+    mixins.UpdateModelMixin,
+):
+    """
+    ViewSet for SharedAttributeManager.
+
+    NOTE: This API endpoint is in "tech preview" and subject to change
+    """
+
+    queryset = SharedAttributeManager.objects.all()
+    endpoint_name = "shared-attribute-managers"
+    filterset_class = SharedAttributeManagerFilter
+    serializer_class = SharedAttributeManagerSerializer
+    ordering = "-pulp_created"
+
+    @extend_schema(
+        description="Update attributes and possibly trigger an asynchronous /apply/ task",
+        responses={200: SharedAttributeManagerSerializer, 202: AsyncOperationResponseSerializer},
+    )
+    def update(self, request, pk, **kwargs):
+        # Update the SAM entity
+        rc = super().update(request, pk, **kwargs)
+        sam = self.get_object()
+        updatable_entities = sam.managed_entities and sam.managed_attributes
+        apply_update = any(
+            key in request.data for key in ("managed_entities", "managed_attributes")
+        )
+        # If we updated entities or attributes, and post-update we have entities *and* attributes,
+        # apply the attributes.
+        if updatable_entities and apply_update:
+            return self.apply(request, pk)
+        else:
+            return rc
+
+    @extend_schema(
+        description="Update partial attributes and possibly trigger an asynchronous /apply/ task",
+        responses={200: SharedAttributeManagerSerializer, 202: AsyncOperationResponseSerializer},
+    )
+    def partial_update(self, request, *args, **kwargs):
+        kwargs["partial"] = True
+        return self.update(request, *args, **kwargs)
+
+    @extend_schema(
+        description="Trigger an asynchronous task to update managed entities.",
+        summary="Update entities",
+        responses={202: AsyncOperationResponseSerializer},
+    )
+    @action(detail=True, methods=["post"])
+    def apply(self, request, pk):
+        """
+        Dispatches a task to apply changes to all managed entities.
+        """
+        sam = self.get_object()
+        if not sam.managed_entities:
+            raise ValidationError(_("No entities are being managed."))
+
+        # Make sure we lock all the entities we may be updating in the "entities" list
+        entities = [sam, *sam.managed_entities]
```

Comment:
:+1:

Reply:
```suggestion
        locks = [sam, *sam.managed_entities]
```

The `sam` itself is in the list and it's not an entity. These are really `locks`.